src/NHibernate/Util/CollectionHelper.cs

```diff
@@ -602,6 +602,38 @@ public static int GetHashCode<T>(IEnumerable<T> coll)
 			}
 		}
 
+		// KeyValuePair<TK, TV>.GetHashCode default implementation is inadequate, see https://stackoverflow.com/a/5927853/1178314
+		// If one of its type is a reference type, it may ignore it for computing the hashcode, leading to potentially
+		// a lot of hashcode collisions. It is not a bug from GetHashCode semantic viewpoint, but it causes them
+		// to be potentially inefficients for usages in dictionnaries or hashset.
+		/// <summary>
+		/// Computes a hash code for <paramref name="coll"/>.
+		/// </summary>
+		/// <remarks>The hash code is computed as the sum of hash codes of
+		/// individual elements key xored with their value if any, so that the value is independent of the
+		/// collection iteration order.
+		/// </remarks>
+		public static int GetHashCode<TK, TV>(IEnumerable<KeyValuePair<TK, TV>> coll)
```

Comment:
The overload added here is a specialization of the existing `public static int GetHashCode<T>(IEnumerable<T> coll)`, allowing to handle the dictionary case specifically without source changes on the `GetHashCode` callers side: with a recompilation, when they supply a dictionary, that is this new overload which gets called, allowing to fix the issue.

Adding comparer overloads would require to change the calling code too (8 cases, in cache, engine and mapping namespaces). I would rather see this as a separate refactoring.

Reply:
But this fix is too specialised. The problem exists for all struct types which can contain references.

Reply:
Can we make this method internal then?

Reply:
Done. But now the test heavily relies on reflection, and it would be better to merge #1612 first then add tests here for checking `QueryKey` and `FilterKey` benefit of the internal overload. (Their tests fixtures are currently broken, causing them to be unusable. #1612 fixes them.)